fasjson_client/client.py

```diff
@@ -0,0 +1,118 @@
+import errno
+from urllib.parse import urlparse
+
+import yaml
+import requests
+from requests_gssapi import HTTPSPNEGOAuth
+from bravado import requests_client
+from bravado.client import SwaggerClient
+import gssapi
+from swagger_spec_validator.common import SwaggerValidationError
+
+from . import errors
+
+
+class HttpClient(requests_client.RequestsClient):
+  """
+  HttpClient class used in SwaggerClient for gssapi authentication.
+  """
+
+  def __init__(self, principal, *args, **kwargs):
+      super().__init__(*args, **kwargs)
+      self.principal = principal
+
+  def authenticated_request(self, request_params):
+    """
+    Retrieves an authentication token from kerberos
+    and adds it in the http header request.
+    """
+    name = gssapi.Name(self.principal, gssapi.NameType.kerberos_principal)
+    try:
+      creds = gssapi.Credentials(name=name, usage="initiate")
+    except gssapi.raw.misc.GSSError as e:
+      data = {
+        "trace": repr(e.gen_message),
+        "codes": {
+          "maj": e.maj_code,
+          "min": e.min_code,
+          "routine": e.routine_code,
+          "supplementary_code": e.supplementary_code,
+        },
+      }
+      raise errors.ClientError(
+        "schema validation failed", errno.EPROTO, data=data
+      )
+
+    request_params["auth"] = HTTPSPNEGOAuth(creds=creds)
+
+    return super().authenticated_request(request_params)
+
+
+class Client:
+  """
+  FasJsonClient client class that builds API methods based on openapi specs.
+  """
+
+  def __init__(self, spec, principal):
+    self.principal = principal
+    self.http_client = HttpClient(principal)
+    try:
+      self._api = SwaggerClient.from_spec(
+        spec,
+        http_client=self.http_client,
+        config={"disable_fallback_results": True},
+      )
+    except SwaggerValidationError:
+      raise errors.ClientError("schema validation failed", errno.EPROTO)
+
+  @classmethod
+  def from_url(cls, spec_url, base_url=None, principal=None):
```

Comment:
Let's go like this for now, but eventually I'd prefer the spec URL to be deducted from the base URL (maybe a /spec endpoint? Or a header? Is there a standard for that?)

Reply:
spec_url is where bravado loads the spec from including the base url whcih is documented in the api spec file, `base_url` is here in case a user wants do overwrite the value used within the spec.

Reply:
we can have a default value for the spec url but I think we need to push fasjson somewhere first?

Reply:
If we decide that fasjson publishes the spec at `/spec` for example (again, if there's a standard location for that, let's use it), then we can just provide a base URL and the client will know where to find the spec.

Reply:
I created an issue to provide it in the root version path, such as `/v1`, so I think we can update it once we have it hosted somewhere/defined a production baseurl.

Reply:
Has it been merged yet? If not could you link to the issue here so we can track it?

Reply:
I created an issue but didn't worked on it - it needs to implement some kind of "white list" which enables us to choose which path should not be authenticated against ldap.

We could set the url from oour git repository while for now if that's ok.

Reply:
OK, let's go that way for now.